display_list/dl_builder.h

```diff
@@ -593,21 +587,23 @@ class DisplayListBuilder final : public virtual DlCanvas,
 
    private:
     size_t save_offset_;
-    bool has_layer_;
+    bool is_save_layer_ = false;
     bool cannot_inherit_opacity_ = false;
     bool has_compatible_op_ = false;
     std::shared_ptr<const DlImageFilter> filter_;
     bool is_unbounded_ = false;
     bool has_deferred_save_op_ = false;
     bool is_nop_ = false;
     bool affects_transparent_layer_ = false;
+    std::shared_ptr<BoundsAccumulator> layer_accumulator_;
```

Comment:
Can we avoid heap allocating another structure here? Could this be a struct on the DLBuilder?

Reply:
For this specific object, it can be null for Save entries, it will be non-null for SaveLayer entries (has_layer) and for Save entries inside a SaveLayer, they need to forward the existing accumulator, so it needs to be shareable. I could do some finagling to try to do some forwarding without having to do a shared pointer, but how much would it save?

Reply:
There are already a lot of heap allocated structures created during recording/replay of the dl and layer trees, for example all of the LSS allocations show up in profiles. Though the overall time is small, it will accumulate  and I'm worried that will eventually become too difficult to fix without a large refactor.

Reply:
I've been accumulating ideas for simplification of a lot of this while doing this PR. The changes don't directly contribute towards eliminating the complexity/existance of the entity structures, but they will be important steps going forward. These would include:

- MatrixClipTracker (DlMCT) has 2 variants for SkMatrix and SkM44 for faster performance for 99% of use cases. I'd like to move to just one transform that is both capable of a full 4x4 transform and has optimizations for the common cases. Actually, just simplifying to SkM44 may be enough in the short term because using it with asM33() to do bounds transforms in order to leverage the optimizations in SkMatrix has a reasonable overhead compared to the performance of impeller::Matrix which does a pessimal operation every time it does a bounds transform. But, moving to impeller::Matrix would be more forward-thinking in terms of eventually weaning the DL off of Skia sources. That's an ongoing background investigation at this point. It's half of the reason why I want to create an impeller::Transform which wraps an impeller::Matrix and tracks the state for optimizations. The state tracking only has minor improvement for things like scale/translate/transform on the matrix itself, but the bounds transform operation using Matrix is way slower than what SkMatrix does.
- If DlMCT is greatly simplified, then it can be split into a singular MatrixClipObject which just stores the matrix and clip and does not provide stack ops, and DlMCT becomes simply a stack of MCOs. Then I can absorb that new object into the existing stack data within DLBuilder and reduce the complexity of needing multiple objects for Save/Restore operations and inline their allocations. DlMCT would still be available for other code uses that don't already have their own stack structures.
- LSS could be re-architected into something stack allocated during Preroll, it would take some work, and not be as "one stop" as the existing code. I originally had something like that, but the complexities of maintaining the linkage between stack allocated structures became tricky for the *Layer classes to manage and hiding it inside a single object with save/restore semantics was simpler to use in the Layers. If it is really important, though, I could move back to a stack allocated variant. The complexities are that LSS supports dynamically changing its "target" (i.e. delegate) on the fly and also to dynamically produce a set of Mutation objects for the Platform Views - all of which involve resolving/replaying the list of outstanding states. Sure, I can have them back-chain through the stack-allocated objects, but it gets tricky.

Reply:
This seems like a reasonable strategy, and moving towards impeller::Matrix long term is a good idea. In general though, I'm skeptical that we'll get much juice out of differentiating between affine and 4x4 matrices - keeping one path might be simpler. And if necessary, more amenable to SIMD accelleration instead.